src/typeChecker/typeErrorChecker.ts

```diff
@@ -0,0 +1,856 @@
+ */
+function handleImportDeclarations(
+  node: NodeWithTypeAnnotation<es.Program>,
+  context: Context,
+  env: TypeEnvironment
+) {
+  const importStmts: es.ImportDeclaration[] = node.body.filter(
+    stmt => stmt.type === 'ImportDeclaration'
+  ) as es.ImportDeclaration[]
+  if (importStmts.length > 0) {
+    const modules = memoizedGetModuleManifest()
+    const moduleList = Object.keys(modules)
+    importStmts.forEach(stmt => {
+      const moduleName = stmt.source.value as string
+      if (!moduleList.includes(moduleName)) {
+        context.errors.push(new ModuleNotFoundError(moduleName, stmt))
+      }
+      stmt.specifiers.map(spec => {
+        if (spec.type !== 'ImportSpecifier') {
+          throw new TypeError(stmt, 'Unknown specifier type')
+        }
+
+        setType(spec.local.name, tAny, env)
+      })
+    })
+  }
+}
+
+/**
+ * Adds all types for variable/function/type declarations to the current environment.
+ * This is so that the types can be referenced before the declarations are initialized.
+ */
+function addTypeDeclarationsToEnvironment(
+  node: es.Program | es.BlockStatement,
+  context: Context,
+  env: TypeEnvironment
+) {
+  node.body.forEach(node => {
+    switch (node.type) {
+      case 'FunctionDeclaration':
+        if (node.id === null) {
+          // Block should not be reached since node.id is only null when function declaration
+          // is part of `export default function`, which is not used in Source
+          throw new TypeError(node, 'Function declaration should always have an identifier')
+        }
+        const params = node.params as NodeWithTypeAnnotation<es.Identifier>[]
+        const returnType = getAnnotatedType(
+          (node as NodeWithTypeAnnotation<es.FunctionDeclaration>).returnType,
+          context,
+          env
+        )
+
+        const types = getParamTypes(params, context, env)
+        // Return type will always be last item in types array
+        types.push(returnType)
+        const fnType = tFunc(...types)
+
+        // Save function type in type env
+        setType(node.id.name, fnType, env)
+        break
+      case 'VariableDeclaration':
+        if (node.kind === 'var') {
+          throw new TypeError(node, 'Variable declaration using "var" is not allowed')
+        }
+        if (node.declarations.length !== 1) {
+          throw new TypeError(node, 'Variable declaration should have one and only one declaration')
+        }
+        if (node.declarations[0].id.type !== 'Identifier') {
+          throw new TypeError(node, 'Variable declaration ID should be an identifier')
+        }
+        const id = node.declarations[0].id as NodeWithTypeAnnotation<es.Identifier>
+        const expectedType = getAnnotatedType(id.typeAnnotation, context, env)
+
+        // Save variable type and decl kind in type env
+        setType(id.name, expectedType, env)
+        setDeclKind(id.name, node.kind, env)
+      default:
```

Comment:
There is no `break` or `return` statement here, so this case will fallthrough to the default case. Is this intended?

Reply:
Thanks for the catch! Though the return type is the same as default case, it is still best to return the value instead of falling through.